### PR TITLE
Added crypto callback for SHA3.

### PR DESCRIPTION
**Pull request by aidangarske**
# Description

Added crypto callback for SHA3 and extended the test.c tests for it in cryptocb_test.

# Testing 

```
./configure --enable-cryptocb --enable-sha3 CFLAGS="-DDEBUG_CRYPTOCB" --enable-debug
make 
./wolfcrypt/test/testwolfcrypt
```

### CODE REVIEW

**Comment by dgarske** [comment]
Retest this please

**Review by dgarske — APPROVED** [review]

**Review by JacobBarthelmeh — APPROVED** [review]

File: wolfcrypt/src/sha3.c [hunk 1]
```diff
@@ -817,10 +820,15 @@ static int wc_InitSha3(wc_Sha3* sha3, void* heap, int devId)
 #if defined(WOLFSSL_ASYNC_CRYPT) && defined(WC_ASYNC_ENABLE_SHA3)
     ret = wolfAsync_DevCtxInit(&sha3->asyncDev,
                         WOLFSSL_ASYNC_MARKER_SHA3, sha3->heap, devId);
-#else
-    (void)devId;
+#elif defined(WOLF_CRYPTO_CB)
+    sha3->devId = devId;
+    sha3->type = type;
```
Comment (billphipps) [inline wolfcrypt/src/sha3.c:825]: In the other cases (async, sw), the p parameter passed into SHA3_Update and SHA3_Final controls which type (256, 384, 512) of SHA3 is being used.  Consider letting the the processing functions determine the info->hash->type based on that and remove the type parameter/member here.  Comments below.

File: wolfcrypt/src/sha3.c [hunk 2]
```diff
@@ -880,6 +899,17 @@ static int wc_Sha3Final(wc_Sha3* sha3, byte* hash, byte p, byte len)
         return BAD_FUNC_ARG;
     }
 
+#ifdef WOLF_CRYPTO_CB
+    #ifndef WOLF_CRYPTO_CB_FIND
+    if (sha3->devId != INVALID_DEVID)
+    #endif
+    {
+        ret = wc_CryptoCb_Sha3Hash(sha3, sha3->type, NULL, 0, hash);
```
Comment (billphipps) [inline wolfcrypt/src/sha3.c:907]: Consider:
int type = 0;
switch(len) {
case WC_SHA3_224_DIGEST_SIZE: type = WC_HASH_TYPE_SHA3_224; break;
case WC_SHA3_256_DIGEST_SIZE: type = WC_HASH_TYPE_SHA3_256; break;
case WC_SHA3_384_DIGEST_SIZE: type = WC_HASH_TYPE_SHA3_384; break;
case WC_SHA3_512_DIGEST_SIZE: type = WC_HASH_TYPE_SHA3_512; break;
default: return WC_ERROR_BAD_ARG;
}
ret = wc_CryptoCb_Sha3Hash(sha, type, NULL, 0, hash);

File: wolfcrypt/src/sha3.c [hunk 3]
```diff
@@ -845,6 +853,17 @@ static int wc_Sha3Update(wc_Sha3* sha3, const byte* data, word32 len, byte p)
         return 0;
     }
 
+#ifdef WOLF_CRYPTO_CB
+    #ifndef WOLF_CRYPTO_CB_FIND
+    if (sha3->devId != INVALID_DEVID)
+    #endif
+    {
+        ret = wc_CryptoCb_Sha3Hash(sha3, sha3->type, data, len, NULL);
```
Comment (billphipps) [inline wolfcrypt/src/sha3.c:861]: Consider:
int type = 0;
switch(p) {
case WC_SHA3_224_COUNT: type = WC_HASH_TYPE_SHA3_224; break;
case WC_SHA3_256_COUNT: type = WC_HASH_TYPE_SHA3_256; break;
case WC_SHA3_384_COUNT: type = WC_HASH_TYPE_SHA3_384; break;
case WC_SHA3_512_COUNT: type = WC_HASH_TYPE_SHA3_512; break;
default: return WC_ERROR_BAD_ARG;
}
ret = wc_CryptoCb_Sha3Hash(sha, type, NULL, 0, hash);

File: wolfcrypt/src/sha3.c [hunk 4]
```diff
@@ -880,6 +899,17 @@ static int wc_Sha3Final(wc_Sha3* sha3, byte* hash, byte p, byte len)
         return BAD_FUNC_ARG;
     }
 
+#ifdef WOLF_CRYPTO_CB
```
Comment (billphipps) [inline wolfcrypt/src/sha3.c:902]: Remember that SHA3_Final will invoke InitSha3(sha3) at the end which will clear out the devId.  This is a bug in the other SHA implementations as well and needs to fixed.

**Review by billphipps — COMMENTED** [review]
Looks ok, but I'd like to see the existing pattern of using the p (block count) and len to determine the type instead of modifying and adding a member to track it.  Also, there's a bug in InitSha3 that clears the devId, so the end of Final should either invoke the wc_InitSha3(..., sha3->devId) or InitSha3 should be updated to remember the devId.

File: wolfcrypt/src/sha3.c (hunk 4 above)
Comment (aidangarske) [reply wolfcrypt/src/sha3.c:902]: Final calls `InitSha3` which does not touch devId or heap hint. This is not an issue (says David Garske)

Comment (billphipps) [reply wolfcrypt/src/sha3.c:902]: Sweet!  I must have gotten confused that it was memsetting the struct.  great to hear.  Thx

**Pull request merged**